consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,431 @@
+			case p := <-t.listeners.ProposalListener.Listen():
+				t.handleProposal(p)
+			case p := <-t.prevotesCh:
+				t.handlePrevote(p)
+			case p := <-t.listeners.PrevoteListener.Listen():
+				t.handlePrevote(p)
+			case p := <-t.precommitsCh:
+				t.handlePrecommit(p)
+			case p := <-t.listeners.PrecommitListener.Listen():
+				t.handlePrecommit(p)
+			}
+			i++
+		}
+	}()
+}
+
+func (t *Tendermint[V, H, A]) Stop() {
+	fmt.Println("Stopping Tendermint...")
+	close(t.quit)
+	for _, tm := range t.scheduledTms {
+		tm.Stop()
+	}
+	fmt.Println("Waiting fo the Tendermint loop to exit")
+	t.wg.Wait()
+}
+
+func (t *Tendermint[V, H, A]) startRound(r uint) {
+	fmt.Println("Inside start round", r)
+	if r != 0 && r <= t.state.round {
+		return
+	}
+
+	t.state.round = r
+	t.state.step = propose
+
+	t.state.line34Executed = false
+	t.state.line36Executed = false
+	t.state.line47Executed = false
+
+	if p := t.validators.Proposer(t.state.height, r); p == t.nodeAddr {
+		var proposalValue *V
+		if t.state.validValue != nil {
+			proposalValue = t.state.validValue
+		} else {
+			v := t.application.Value()
+			proposalValue = &v
+		}
+		proposalMessage := Proposal[V, H, A]{
+			Height:     t.state.height,
+			Round:      r,
+			ValidRound: t.state.validRound,
+			Value:      proposalValue,
+			Sender:     t.nodeAddr,
+		}
+
+		t.messages.addProposal(proposalMessage)
+		fmt.Println("Broadcasting the proposal")
+		t.broadcasters.ProposalBroadcaster.Broadcast(proposalMessage)
+	} else {
+		t.scheduleTimeout(t.timeoutPropose(r), propose, t.state.height, t.state.round)
+	}
+
+	go t.processFutureMessages(t.state.height, t.state.round)
+}
+
+//nolint:gocyclo
+func (t *Tendermint[V, H, A]) processFutureMessages(h, r uint) {
+	t.futureMessagesMu.Lock()
+	defer t.futureMessagesMu.Unlock()
+
+	proposals, prevotes, precommits := t.futureMessages.allMessages(h, r)
+	if len(proposals) > 0 {
+		for _, addrProposals := range proposals {
+			for _, proposal := range addrProposals {
+				select {
+				case <-t.quit:
+					return
+				case t.proposalsCh <- proposal:
+				}
+			}
+		}
+	}
+
+	if len(prevotes) > 0 {
+		for _, addrPrevotes := range prevotes {
+			for _, vote := range addrPrevotes {
+				select {
+				case <-t.quit:
+					return
+				case t.prevotesCh <- vote:
+				}
+			}
+		}
+	}
+
+	if len(precommits) > 0 {
+		for _, addrPrecommits := range precommits {
+			for _, vote := range addrPrecommits {
+				select {
+				case <-t.quit:
+					return
+				case t.precommitsCh <- vote:
+
+				}
+			}
+		}
+	}
+
+	t.futureMessages.deleteRoundMessages(h, r)
+}
+
+type timeout struct {
+	*time.Timer
+
+	s step
+	h uint
+	r uint
+}
+
+func (t *Tendermint[V, H, A]) scheduleTimeout(duration time.Duration, s step, h, r uint) {
+	tm := timeout{s: s, h: h, r: r}
+	tm.Timer = time.AfterFunc(duration, func() {
+		select {
+		case <-t.quit:
+		case t.timeoutsCh <- tm:
+		}
+	})
+	t.scheduledTms = append(t.scheduledTms, tm)
+}
+
+func (t *Tendermint[_, H, A]) OnTimeoutPropose(h, r uint) {
```

Comment:
So imagine that we start up a node, and it never recieves messages from other nodes. We start round 0, schedule the propose timeout, which eventually gets hit, and we end up here. Then we cast a `nil` prevote, set `step` to `prevote` and continue. We will end up back at `startRound(0)` again, reset everything, and the process will repeat indefinitely, right? We would effectively be casting `nil` prevotes with `round=0` forever? (I would have thought the round would increase..)